kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -216,16 +220,27 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
             completed = true
             continue
         }
-        select<Unit> {
-            coroutine.onJoin {
-                completed = true
-            }
-            scheduler.onDispatchEvent {
-                // we received knowledge that `scheduler` observed a dispatch event, so we reset the timeout
+        // in case progress depends on some background work, we need to keep spinning it.
+        val backgroundWorkRunner = launch(CoroutineName("background work runner")) {
+            while (true) {
+                scheduler.tryRunNextTaskUnless { !isActive }
```

Comment:
Is my understanding of this machinery correct?

* We cannot run background jobs in the previous `advanceUntilIdle` because otherwise, we may be stuck in an endless loop of background work
* `yield` is required to get the chance to `select` to complete and switch `isActive` trigger
* We cannot avoid this `launch` even though we get notified in `onDispatchEvent` because of p.1

Reply:
I've tried to figure out the way to avoid ceremonies with extra `launch`, but it doesn't seem to be that easy.

The machinery itself is already quite complex and it would be nice if you could drop a comment with a rationale behind it